File: Python/Diameter of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution(object):
    def diameterOfBinaryTree(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.result = 0
        self.deepth(root)
        return self.result
        
        
        
    def deepth(self, root):
        if root == None:
            return 0
        else:
            left_deepth = self.deepth(root.left)
            right_deepth = self.deepth(root.right)
            self.result = max(self.result, left_deepth + right_deepth)
            return max(left_deepth, right_deepth) + 1

File: cpp/Diameter of Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int result = 0;
    int diameterOfBinaryTree(TreeNode* root) {
        depth(root);
        return result;
    }

    int depth(TreeNode *root) {
        if (!root)
            return 0;
        else {
            int left_depth = depth(root->left);
            int right_depth = depth(root->right);
            result = max(result, left_depth + right_depth);
            return max(left_depth, right_depth) + 1;
        }
    }

};
